src/reducers/console.js

```diff
@@ -35,20 +34,55 @@ export default function console(stateIn, {type, payload, meta}) {
         ),
       );
     case 'EVALUATE_CONSOLE_ENTRY':
-      return payload.trim(' ') === '' ? state : state.set(
-        meta.key,
+      return payload.trim(' ') === '' ? state : state.setIn(
+        ['history', meta.key],
         new ConsoleEntry({expression: payload}),
-      );
+      ).
+        delete('currentInputValue').
+        delete('nextConsoleEntry').
+        delete('historyEntryIndex');
     case 'CLEAR_CONSOLE_ENTRIES':
       return initialState;
+    case 'CONSOLE_INPUT_CHANGED':
+      return state.set('currentInputValue', payload.value);
     case 'CONSOLE_LOG_PRODUCED':
-      return state.set(
-        meta.key,
+      return state.setIn(
+        ['history', meta.key],
         new ConsoleEntry({
           value: payload.value,
           evaluatedByCompiledProjectKey: payload.compiledProjectKey,
         }),
       );
+    case 'NAVIGATE_CONSOLE_HISTORY': {
+      const newHistoryEntryIndex = payload.direction === 'UP' ?
+        state.historyEntryIndex + 1 :
+        state.historyEntryIndex - 1;
+
+      const expressionHistory = state.history.toList().
+        map(entry => entry.expression).
+        filter(expression => expression !== null).
+        concat(state.nextConsoleEntry).
+        reverse();
```

Comment:
I don’t feel strongly about it, but [`Immutable.List#get()`](https://facebook.github.io/immutable-js/docs/#/List/get) does allow you to index into the end of a list with negative numbers, which would maybe retain the pleasantness while avoiding the need to reverse the list every time.

Reply:
It seems like not a clear win to me, since -0 is 0 so I can't just use negative index.

Reply:
Good call